### DISCUSSION TITLE
Where are the textures of the terrain edges located?

### DISCUSSION THREAD

**Discussion by karst** [OpenLoco/OpenLoco#1592] [Q&A]
Heya,
I was looking around through the ObjData of the game as I am looking for the terrain edge textures.
I was thinking they might be in `ROCK1.DAT` and  `ROCK2.DAT` but these are only the terrain surface objects.
If someone's able to direct me to them it would save me alot of time exporting every DAT file one by one.
Thanks!

**Comment by AaronVanGeffen** [comment]
The paint routines for surfaces (or edges) haven't been reimplemented in C++ yet. Not sure how it works yet. Perhaps @duncanspumpkin has done some exploratory work on this, though?

**Comment by duncanspumpkin** [comment]
I've not looked into it. It looked like the code was very similar to rct2 though so could probably find out without much bother. I won't be able to look into this until next week ping me a message on Thursday because I'll probably forget.

**Comment by karst** [comment]
It's thursday! @duncanspumpkin

**Comment by duncanspumpkin** [comment]
okay so i think https://github.com/OpenRCT2/OpenRCT2/blob/bbd59511541d29bd390f9075d688854138bdbebe/src/openrct2/paint/tile_element/Paint.Surface.cpp#L1308 is `0x46655F` which is calling `0x00464742`.
Within that we have all of the drawing (which should be the edges) use base image index from `0xF25314` which looks to be set by `LandObject->var_12`. `LandObject->var_12` is loaded from `RockObject->image`. So to me it looks like RockObject is the one that has the edges. The actual image offset within RockObject that is the edge I could work out but it would take a while to decipher.

**Comment by karst** [comment]
this is so odd. The texture on the sides look like this.
![image](https://user-images.githubusercontent.com/1516893/184196179-5fcb0347-ad81-4adf-9c1b-cc5bb09fdc8e.png)

But the image sheet that OpenRCT2 is printing from the DAT file is this:
![images](https://user-images.githubusercontent.com/1516893/184196297-3b9f0535-8e3f-4637-b13d-3ddf538f39d2.png)

I do not see the texture of the side anywhere in here.

**Reply by IntelOrca** [reply]
The vertical textures in Loco are unique to Loco, none of them were in RCT2, or not without changes. In Loco, you get a texture which can wrap across 4 to 8 height levels? Where as in RCT2, it was limited to wrapping every 1 height level.

**Reply by karst** [reply]
I see, would be neat to see that functionality in OpenRCT2. The rock textures are sought after by the NE community so I am trying to see if I can get it to work as custom element there.

**Reply by IntelOrca** [reply]
I would like to see OpenRCT2 add the improved vertical face rendering from Loco, but I don't know what needs to be done. We might need to wait for that part to be implemented first in OpenLoco. @duncanspumpkin any thoughts?

**Reply by duncanspumpkin** [reply]
Ahh looks like it uses the mask code to do the drawing. So it uses g1 images around 3731 to mask the vertical face against the rock texture. So it works a good bit differently compared to RCT2.

**Reply by AaronVanGeffen** [reply]
Ahh, the good old times when we thought the mask code was only used for tree wilting... 😅

**Reply by duncanspumpkin** [reply]
Note this is the real mask code not the wilt code (renamed to noise code) which is subtly a different thing.
